packages/opentelemetry-tracer/src/instrumentation/PluginLoader.ts

```diff
@@ -0,0 +1,148 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Tracer, Logger, Plugin } from '@opentelemetry/types';
+import * as utils from './utils';
+import * as hook from 'require-in-the-middle';
+
+// States for the Plugin Loader
+export enum HookState {
+  UNINITIALIZED,
+  ENABLED,
+  DISABLED,
+}
+
+interface PluginNames {
+  [pluginName: string]: string;
+}
+
+// tslint:disable-next-line:no-any
+export type PluginType = Array<Plugin<any>>;
+
+/**
+ * The PluginLoader class can load instrumentation plugins that
+ * use a patch mechanism to enable automatic tracing for
+ * specific target modules.
+ */
+export class PluginLoader {
+  /** A list of loaded plugins. */
+  private _plugins: PluginType = [];
+  /**
+   * A field that tracks whether the require-in-the-middle hook has been loaded
+   * for the first time, as well as whether the hook body is activated or not.
+   */
+  private _hookState = HookState.UNINITIALIZED;
+
+  /** Constructs a new PluginLoader instance. */
+  constructor(readonly tracer: Tracer, readonly logger: Logger) {}
+
+  get state(): HookState {
+    return this._hookState;
+  }
+
+  get plugins(): PluginType {
+    return this._plugins;
+  }
+
+  /**
+   * Loads a list of plugins. Each plugin module should implement the core
+   * {@link Plugin} interface and export an instance named as 'plugin'. This
+   * function will attach a hook to be called the first time the module is
+   * loaded.
+   * @param modulesToPatch A list of plugins.
+   */
+  loadPlugins(modulesToPatch: string[]): PluginLoader {
```

Comment:
How can plugins be configured?

Reply:
Also, how can custom plugins be used?

Reply:
> How can plugins be configured?

I was thinking to use list of strings to configure avaiable plugins. Something like,
`const DEFAULT_INSTRUMENTATION_PLUGINS = ['http', 'express', 'mysql', ...]` and use this const when calling the PluginLoader inside node-tracer constructor. WDYT?

Reply:
What I meant is, how can configuration options be passed to individual plugins?

One way to do it would be to accept an object with plugin names as keys as configurations as values. To make this easier to use when configuration is not necessary, a boolean could be accepted as well.

For example:

**Enable plugin manually**

```js
loadPlugins({
  express: true
})
```

**Disable plugin manually**

```js
loadPlugins({
  express: false
})
```

**Configure plugin manually (also implicitly enables it)**

```js
loadPlugins({
  express: {
    middleware: false // configures Express to not have spans per middleware (example)
  }
})
```